### ISSUE TITLE
@typescript-eslint can't parse TypeScript 5.x tsconfig.json file format when using PNPM

### ISSUE THREAD

**Issue by matthew-dean** [typescript-eslint/typescript-eslint#7499]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Repro Code

```typescript
npm eslint --config .eslintrc.cjs .
// or 
yarn eslint --config .eslintrc.cjs .
// or
pnpm eslint --config .eslintrc.cjs .
```
### ESLint Config

module.exports = {
  parser: '@typescript-eslint/parser',
  extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended'],
}
```
```

### tsconfig

```jsonc
{
  "extends": ["tsconfig.json"]
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Since https://github.com/microsoft/TypeScript/issues/29118, ESLint should also be able to parse it and run normally.

### Actual Result

Parsing error: Compiler option 'extends' requires a value of type string

### Additional Info

This is a dupe of https://github.com/typescript-eslint/typescript-eslint/issues/6219, but it was closed without investigation, and it cannot be commented on. I also had to check `I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.` because otherwise you can't submit an issue.

Work publicly started on this on or around [January 20, 2021](https://github.com/microsoft/TypeScript/issues/29118#event-4230606938). This was announced as implemented in [January 2023](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#supporting-multiple-configuration-files-in-extends) and released on [March 16, 2023.](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.4.0` |
| `@typescript-eslint/parser`            | `6.4.0` |
| `TypeScript`                           | `5.0.3` |
| `ESLint`                               | `8.47.0` |
| `node`                                 | `16.19.0` |

**Comment by JoshuaKGoldberg** [comment]
This was noted in #6380 and just never acted on because nobody's asked for it yet. I'm honestly surprised it took this long for someone to file an issue post-5.0 😄. Accepting PRs!

**Comment by bradzacher** [comment]
I don't think that we manually parse tsconfigs - we pass them through to TS for parsing (which is why the 5.0 notes don't mention anything for us to do). I'm surprised that this errors.

Part of me thinks there might be a version mismatch.

@matthew-dean can you provide an isolated repro repo showing the error?

**Comment by Zamiell** [comment]
Just chiming in to say that I use tsconfig `extends` with the array form in my monorepo in combination with most of the rules from this project, and I don't have any issues. So it sounds like this issue might be probably something specific to his ESLint config.

**Comment by bradzacher** [comment]
A long time ago in a galaxy far, far away the first version of our tool was built with `typescript: *`. It purposely did not include a version range because TS versions are funky to begin with, and the tools supported all versions anyways. We also didn't want to restrict people to just the versions we declared because by-and-large our tool would work with new versions too - but we often lagged behind the releases and thus didn't want to hold users back. So a runtime warning was added which told people to beware if they were doing something we hadn't tested - but otherwise gave the user free reign.

Some time after that the ecosystem started to build tooling on top of our tooling (eg create-react-app). The problem was that at the time we specified a peer dependency range of `typescript: *` which meant that users of such tools who did not use TS would get stuck with console logs at install time warning them they were missing a package. This warning was wrong as it was all optional but it annoyed people and they loudly complained at the tool owners and at us.

Our solution was to just remove the peer dependency specification. We already had runtime validation of the version and the entire system would break without it - so it was essentially superfluous.

Some time after that yarn2 introduced a feature that would enforce that packages do not access any other package unless it was explicitly declared as a dependency.

We, of course, failed this assertion and people could not use our tooling under new yarn. The workaround was for us to add this config to our packages which would tell package managers we did actually depend on TS but it was okay if it didn't exist.

https://github.com/typescript-eslint/typescript-eslint/blob/4f34d0ba34474926ba1eed623704b583a037f886/packages/eslint-plugin/package.json#L97-L101

This satisfied the constraints and made the ecosystem work and everyone was happy: users didn't get any warning logs on any version of any package manager because we didn't have `typescript: *`, and new package managers assumed we meant that thanks to `peerDependenciesMeta` and "just worked".

That's the history of why we do what we do and it's setup in this exact manner. It's been working for a good number of years - so if it's breaking now then that's a regression in pnpm that will need to be solved in some way or another.

**Comment by bradzacher** [comment]
Sorry I was just dumping the entire history here as well so you can share it with the pnpm maintainer if needs be :)

I'm not saying we're infallible here btw - our setup unique and unconventional, even if it has its reasons for existing!

There might be something we can do to fix things - we're completely open to making changes as long as we don't cause logs for the CRA case or restrict users from new TS releases.

Last time we made a change it was the yarn lead maintainer telling us exactly how to fix things - so we would likely want a similarly knowledgeable individual to advise again :)

**Comment by zkochan** [comment]
This statement is not correct:

> PNPM reasons that peer dependencies are related to the package that lists them as peers. And for whatever reason, in their great wisdom, the PNPM decided that, by default, peers that are do not match are auto-installed.

pnpm will resolve a peer dependency using any package that is found in the dependencies of a direct or indirect parent package. It doesn't matter if the found package doesn't match the range in the `peerDependencies` field. A warning will be printed but the package will not be auto-installed as it is found.

**Comment by bradzacher** [comment]
If zoltan says it works then I'm inclined to believe that :)

 it's probably local setup problems for users like outdated versions or something?

Either way I don't think there's anything we should action in our project - so closing this off.

**Comment by zkochan** [comment]
In any case, if you can create a sample repository that reproduces the issue, feel free to create a new issue in the pnpm repository and we'll look into it.

**Comment by zkochan** [comment]
That is a possibility. If you can submit an issue with steps to repro someone can check.
